### PR TITLE
Adding Training/Evaluation/Visualization script and Notebooks

### PR DESCRIPTION
**Pull request by bhattg**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,7 +1,145 @@
-# MRL
+# [Matryoshka Representations for Adaptive Deployment](https://arxiv.org/abs/2205.13147)
+Aditya Kusupati, Gantavya Bhatt, Aniket Rege, Matthew Wallingford, Aditya Sinha, Vivek Ramanujan, William Howard-Snyder, Kaifeng Chen, Sham Kakade, Prateek Jain, Ali Farhadi
 
-This project is primarily organized into 
-1. Training MRL models
-2. Evaluating MRL models via image retrieval, few shot learning
-3. Analysis of MRL models via superclass hierarchy and oracle performance
+This repository contains the code for training and evaluating Matryoshka Representation with CNN backbone, along with visualizations. Training code is built upon [FFCV](https://github.com/libffcv/ffcv-imagenet) modified for MRL. The repository is organized as follows:
```
Comment (aniketrege) [inline README.md]: replace CNN with ResNet-50? We did not use any other backbone (ie mobilenet) in the paper right?

File: README.md [hunk 2]
```diff
@@ -1,7 +1,145 @@
...
+5. Evaluating pre-trained checkpoints
+5. Retrieval Performance
+
+Lastly, we also provide code to generate the figures mentioned in our paper. Please have a look at thee `Visualization/` folder. 
+## Set Up
+First pip install the requirements file in this directory:
+```
+pip install -r requirements.txt
+```
+We use PyTorch Distributed Data Parallel shared over 2 A100 GPUs and FFCV for the dataloader. Please make sure to scale the learning rate according to the #GPUs used.
```
Comment (aniketrege) [inline README.md]: We can write that we linearly downscaled the original learning rate by 4 as we had 4x less GPU than original FFCV benchmark

File: README.md [hunk 3]
```diff
@@ -1,7 +1,145 @@
...
+
+## Training ResNet50 models
+First of all change the directory to `train_and_eval/train`. While training, we dump  model ckpts, training logs and run command inside a folder. 
+
+### Training Fixed Feature Baseline
+
+```bash 
+export CUDA_VISIBLE_DEVICES=0,1,2,3
+
+python train_imagenet.py --config-file rn50_configs/rn50_40_epochs.yaml --model.fixed_feature=2048 --data.train_dataset=/home/jupyter/dataset-ffcv/train_500_0.50_90.ffcv --data.val_dataset=/home/jupyter/dataset-ffcv/uncompressed/val_500_uncompressed.ffcv --data.num_workers=12 --data.in_memory=1 --logging.folder=trainlogs --logging.log_level=1 --dist.world_size=2 --training.distributed=1 --lr.lr=0.425
```
Comment (aniketrege) [inline README.md]: replace `/home/jupyter/dataset-ffcv/` and `/home/jupyter/dataset-ffcv/uncompressed/` with `WRITE_DIR`

File: README.md [hunk 4]
```diff
@@ -1,7 +1,145 @@
...
+```
+
+### Training MRL model
+
+```bash 
+export CUDA_VISIBLE_DEVICES=0,1,2,3
+
+python train_imagenet.py --config-file rn50_configs/rn50_40_epochs.yaml --model.mrl=1 --data.train_dataset=/home/jupyter/dataset-ffcv/train_500_0.50_90.ffcv --data.val_dataset=/home/jupyter/dataset-ffcv/uncompressed/val_500_uncompressed.ffcv --data.num_workers=12 --data.in_memory=1 --logging.folder=trainlogs --logging.log_level=1 --dist.world_size=2 --training.distributed=1 --lr.lr=0.425
```
Comment (aniketrege) [inline README.md]: replace `/home/jupyter/dataset-ffcv/` and `/home/jupyter/dataset-ffcv/uncompressed/` with `WRITE_DIR`

File: README.md [hunk 5]
```diff
@@ -1,7 +1,145 @@
...
+
+By default, we start nesting from 8 dimensions (That is 2^3). In case we want to nest representations from 16 dimensions, add an additional flag of 
+```
+--model.nesting_start=4
+```
+
+### Training MRL-E model
+
+```bash 
+export CUDA_VISIBLE_DEVICES=0,1,2,3,4
```
Comment (aniketrege) [inline README.md]: We used only two to train, so just 0,1 right? we have passed `world_size=2` below as well

File: README.md [hunk 6]
```diff
@@ -1,7 +1,145 @@
...
+If you find this project useful in your research, please consider citing:
+```
+@article{Kusupati2022MatryoshkaRF,
+  title={Matryoshka Representations for Adaptive Deployment},
+  author={Aditya Kusupati and Gantavya Bhatt and Aniket Rege and Matthew Wallingford and Aditya Sinha and Vivek Ramanujan and William Howard-Snyder and Kaifeng Chen and Sham M. Kakade and Prateek Jain and Ali Farhadi},
+  journal={ArXiv},
+  year={2022},
+  volume={abs/2205.13147}
+}
+```
```
Comment (aniketrege) [inline README.md:125]: remove all `.DS_Store `files and `__pycache__ `directories

File: Visualization/Adaptive Retrieval/PLOTS - Sheet15.csv [hunk 7]
```diff
@@ -0,0 +1,7 @@
+FLOPS,MAP@10
```
Comment (aniketrege) [inline Visualization/Adaptive Retrieval/PLOTS - Sheet15.csv]: @adityakusupati do we need to provide all these CSV files and scripts here?

File: Visualization/Adaptive Retrieval/.ipynb_checkpoints/4M-checkpoint.ipynb [hunk 8]
```diff
@@ -0,0 +1,358 @@
+{
```
Comment (aniketrege) [inline Visualization/Adaptive Retrieval/.ipynb_checkpoints/4M-checkpoint.ipynb]: checkpoints can be deleted

File: Visualization/Adaptive Retrieval/Untitled.ipynb [hunk 9]
```diff
@@ -0,0 +1,271 @@
+{
```
Comment (aniketrege) [inline Visualization/Adaptive Retrieval/Untitled.ipynb]: needs to be renamed, along with pdf file below

File: Visualization/Interpolation/.ipynb_checkpoints/Untitled-checkpoint.ipynb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (aniketrege) [inline Visualization/Interpolation/.ipynb_checkpoints/Untitled-checkpoint.ipynb]: delete

File: Visualization/Interpolation/Untitled.ipynb [hunk 11]
```diff
@@ -0,0 +1,660 @@
+{
```
Comment (aniketrege) [inline Visualization/Interpolation/Untitled.ipynb]: rename

File: Visualization/MAP@10/.ipynb_checkpoints/Untitled-checkpoint.ipynb [hunk 12]
```diff
@@ -0,0 +1,270 @@
+{
```
Comment (aniketrege) [inline Visualization/MAP@10/.ipynb_checkpoints/Untitled-checkpoint.ipynb]: delete

File: Visualization/MAP@10/Untitled.ipynb [hunk 13]
```diff
@@ -0,0 +1,286 @@
+{
```
Comment (aniketrege) [inline Visualization/MAP@10/Untitled.ipynb]: rename

File: train_and_eval/train/train_imagenet.py [hunk 14]
```diff
@@ -0,0 +1,607 @@
+import torch as ch
```
Comment (aniketrege) [inline train_and_eval/train/train_imagenet.py]: delete `train.py` which was a template and currently empty file

Comment (aniketrege) [inline train_and_eval/train/train_imagenet.py]: We should write some header comment saying this code was taken directly from FFCV repo? @adityakusupati

File: train_and_eval/train/train_imagenet.py [hunk 15]
```diff
@@ -0,0 +1,607 @@
...
+    @param('training.distributed')
+    @param('model.efficient')
+    @param('model.mrl')
+    @param('model.nesting_start')
+    @param('model.fixed_feature')
+    def __init__(self, gpu, distributed, efficient, mrl, nesting_start, fixed_feature):
+        self.all_params = get_current_config(); 
+        self.gpu = gpu
+        self.efficient = efficient
+        self.mrl = mrl
```
Comment (aniketrege) [inline train_and_eval/train/train_imagenet.py]: delete: `self.mrl` is not being used anywhere besides to set `self.nesting`

File: train_and_eval/train/train_imagenet.py [hunk 16]
```diff
@@ -0,0 +1,607 @@
...
+)
+
+Section('resolution', 'resolution scheduling').params(
+    min_res=Param(int, 'the minimum (starting) resolution', default=160),
+    max_res=Param(int, 'the maximum (starting) resolution', default=160),
+    end_ramp=Param(int, 'when to stop interpolating resolution', default=0),
+    start_ramp=Param(int, 'when to start interpolating resolution', default=0)
+)
+
+Section('data', 'data related stuff').params(
```
Comment (aniketrege) [inline train_and_eval/train/train_imagenet.py]: remove all "stuff" words in Section descriptions and just write "data parameters"

File: train_and_eval/train/train_imagenet.py [hunk 17]
```diff
@@ -0,0 +1,607 @@
...
+
+Section('lr', 'lr scheduling').params(
+    step_ratio=Param(float, 'learning rate step ratio', default=0.1),
+    step_length=Param(int, 'learning rate step length', default=30),
+    lr_schedule_type=Param(OneOf(['step', 'cyclic', 'constant']), default='cyclic'),
+    lr=Param(float, 'learning rate', default=0.5),
+    lr_peak_epoch=Param(int, 'Epoch at which LR peaks', default=2),
+)
+
+Section('logging', 'how to log stuff').params(
```
Comment (aniketrege) [inline train_and_eval/train/train_imagenet.py]: logging parameters

File: train_and_eval/train/train_imagenet.py [hunk 18]
```diff
@@ -0,0 +1,607 @@
...
+    folder=Param(str, 'log location', required=True),
+    log_level=Param(int, '0 if only at end 1 otherwise', default=1)
+)
+
+Section('validation', 'Validation parameters stuff').params(
```
Comment (aniketrege) [inline train_and_eval/train/train_imagenet.py]: remove "stuff" everywhere

File: train_and_eval/train/train_imagenet.py [hunk 19]
```diff
@@ -0,0 +1,607 @@
...
+    @param('model.efficient')
+    @param('model.mrl')
+    @param('model.nesting_start')
+    @param('model.fixed_feature')
+    def __init__(self, gpu, distributed, efficient, mrl, nesting_start, fixed_feature):
+        self.all_params = get_current_config(); 
+        self.gpu = gpu
+        self.efficient = efficient
+        self.mrl = mrl
+        self.nesting = (self.efficient or self.mrl)
```
Comment (aniketrege) [inline train_and_eval/train/train_imagenet.py]: `self.nesting = self.efficient or mrl`

File: train_and_eval/eval/calibration_tools.py [hunk 20]
```diff
@@ -0,0 +1,202 @@
+import numpy as np
```
Comment (aniketrege) [inline train_and_eval/eval/calibration_tools.py]: header comment for which part of paper this code was used for (and for all files in eval/ dir)

File: train_and_eval/eval/eval_ff.py [hunk 21]
```diff
@@ -0,0 +1,111 @@
+import torch
```
Comment (aniketrege) [inline train_and_eval/eval/eval_ff.py]: Header comment description for what this code was used for

File: train_and_eval/train/MRL.py [hunk 22]
```diff
@@ -0,0 +1,52 @@
+import torch
```
Comment (aniketrege) [inline train_and_eval/train/MRL.py]: Small header comment description

File: train_and_eval/train/train_imagenet.py [hunk 23]
```diff
@@ -0,0 +1,607 @@
...
+                    loss_val = self.loss(output, target)
+                    self.val_meters['loss'](loss_val)
+
+        stats = {k: m.compute().item() for k, m in self.val_meters.items()}
+        [meter.reset() for meter in self.val_meters.values()]
+        return stats
+
+
+    @param('validation.lr_tta')
+    def val_loop_nesting(self, lr_tta):
```
Comment (aniketrege) [inline train_and_eval/train/train_imagenet.py]: can we combine this method with `val_loop()` and have an internal flag to reduce code repetition?

File: train_and_eval/train/write_imagenet.py [hunk 24]
```diff
@@ -0,0 +1,66 @@
+from torch.utils.data import Subset
```
Comment (aniketrege) [inline train_and_eval/train/write_imagenet.py]: header that this was taken from FFCV repo

File: train_and_eval/train/write_imagenet.sh [hunk 25]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
```
Comment (aniketrege) [inline train_and_eval/train/write_imagenet.sh]: header that this was taken from FFCV repo

**Review by aniketrege — COMMENTED** [review]
Fixes needed

File: Visualization/Adaptive Retrieval/PLOTS - Sheet15.csv (hunk 7 above)
Comment (adityakusupati) [reply Visualization/Adaptive Retrieval/PLOTS - Sheet15.csv]: I don't mind but if we want to we need to clean up the names, PLOTS - Sheet blah doesn't work. It is generally not needed, but if someone likes our plots, they can reuse and add to our plots.

File: Visualization/Interpolation/.ipynb_checkpoints/Untitled-checkpoint.ipynb (hunk 10 above)
Comment (adityakusupati) [reply Visualization/Interpolation/.ipynb_checkpoints/Untitled-checkpoint.ipynb]: Rename files, this naming is hard to follow.

File: eval_ckpts/autoslim.py [hunk 26]
```diff
@@ -0,0 +1,393 @@
+import torch.nn as nn
```
Comment (adityakusupati) [inline eval_ckpts/autoslim.py]: We don't need baselines.

File: eval_ckpts/eval_ff.py [hunk 27]
```diff
@@ -0,0 +1,111 @@
+import torch
```
Comment (adityakusupati) [inline eval_ckpts/eval_ff.py]: what is the difference between both these files, eval_MRL.py and this one?

File: eval_ckpts/eval_slimmable_network.py [hunk 28]
```diff
@@ -0,0 +1,119 @@
+import torch
```
Comment (adityakusupati) [inline eval_ckpts/eval_slimmable_network.py]: delete

File: train_and_eval/eval/calibration_tools.py (hunk 20 above)
Comment (adityakusupati) [reply train_and_eval/eval/calibration_tools.py]: yes.

File: README.md (hunk 1 above)
Comment (bhattg) [reply README.md]: Replacing it.

File: README.md (hunk 2 above)
Comment (bhattg) [reply README.md]: Incorporating.

File: README.md (hunk 5 above)
Comment (bhattg) [reply README.md]: Yeah, we did. However, I am unsure if that threw us some error since we have os.cache=1. I can change that though.

File: README.md (hunk 3 above)
Comment (bhattg) [reply README.md]: Good catch

File: train_and_eval/train/train_imagenet.py (hunk 16 above)
Comment (bhattg) [reply train_and_eval/train/train_imagenet.py]: That is directly taken from FFCV.

File: train_and_eval/train/train_imagenet.py (hunk 15 above)
Comment (bhattg) [reply train_and_eval/train/train_imagenet.py]: Noted.

File: train_and_eval/train/train_imagenet.py (hunk 14 above)
Comment (bhattg) [reply train_and_eval/train/train_imagenet.py]: Added.

File: train_and_eval/train/train_imagenet.py (hunk 19 above)
Comment (bhattg) [reply train_and_eval/train/train_imagenet.py]: Noted.

File: train_and_eval/eval/calibration_tools.py (hunk 20 above)
Comment (bhattg) [reply train_and_eval/eval/calibration_tools.py]: This is actually never used in the main paper, however it provides a provision to evaluate on ImageNet-O

File: train_and_eval/train/train_imagenet.py (hunk 23 above)
Comment (bhattg) [reply train_and_eval/train/train_imagenet.py]: Ah it would get messy... there isn't actually much of code repeatition.

Comment (aniketrege) [reply train_and_eval/train/train_imagenet.py]: no worries. relatively minor

File: train_and_eval/eval/eval_MRL.py [hunk 29]
```diff
@@ -49,60 +52,32 @@
 elif args.dataset == 'A':
```
Comment (aniketrege) [inline train_and_eval/eval/eval_MRL.py]: We should perhaps add links to these datasets or relevant papers in the header, or describe them with little more verbosity

File: train_and_eval/eval/eval_MRL.py [hunk 30]
```diff
@@ -49,60 +52,32 @@
...
-elif args.dataset == 'C':
-	print("Not implemented Yet~"); sys.exit()
 elif args.dataset == 'R':
 	print("Loading true Imagenet-R val set")
 	dataset = torchvision.datasets.ImageFolder('imagenet-r_/', transform=t)
 elif args.dataset == 'sketch':
 	print("Loading Imagenet-Sketch dataset")
 	dataset = torchvision.datasets.ImageFolder('sketch/', transform=t)
-elif args.dataset == 'O':
-	print("Loading true Imagenet-O val set")
```
Comment (aniketrege) [inline train_and_eval/eval/eval_MRL.py]: This visually looks like the number 0 and not O, is this correct?

Comment (bhattg) [reply train_and_eval/eval/eval_MRL.py]: This should've been removed actually. We decided to remove O altogether.

File: train_and_eval/eval/eval_MRL.py (hunk 29 above)
Comment (bhattg) [reply train_and_eval/eval/eval_MRL.py]: I've added V2 in requirements.txt 
For others, I will add that in the readme.

File: train_and_eval/train/MRL.py [hunk 31]
```diff
@@ -0,0 +1,72 @@
...
```
Comment (adityakusupati) [inline train_and_eval/train/MRL.py]: Why do we have 2 MRL.py. Why isn't it outside? We do not need them in two folders.

Comment (adityakusupati) [reply train_and_eval/train/MRL.py]: It can be in utils. Also is there a reason we need train and eval folders? can't we just have only one? 

I like two, but probably we don't need two folders.

File: eval_ckpts/NestingLayer.py [hunk 32]
```diff
@@ -0,0 +1,77 @@
+import torch
```
Comment (adityakusupati) [inline eval_ckpts/NestingLayer.py]: Can we rename this. If we use nesting in the title, or classes, it would be confusing, can we rename them to something else?

File: eval_ckpts/eval_MRL.py [hunk 33]
```diff
@@ -0,0 +1,84 @@
+import torch
```
Comment (adityakusupati) [inline eval_ckpts/eval_MRL.py]: Why is the file naming not consistent across all the .py files?

File: README.md [hunk 34]
```diff
@@ -1,7 +1,123 @@
...
+
+## Training ResNet50 models
+First of all change the directory to `train_and_eval/train`. While training, we dump  model ckpts, training logs and run command inside a folder. 
+
+### Training Fixed Feature Baseline
+
+```bash 
+export CUDA_VISIBLE_DEVICES=0,1
+
+python train_imagenet.py --config-file rn50_configs/rn50_40_epochs.yaml --model.fixed_feature=2048 --data.train_dataset=$WRITE_DIR/train_500_0.50_90.ffcv --data.val_dataset=$WRITE_DIR/val_500_uncompressed.ffcv --data.num_workers=12 --data.in_memory=1 --logging.folder=trainlogs --logging.log_level=1 --dist.world_size=2 --training.distributed=1 --lr.lr=0.425
```
Comment (adityakusupati) [inline README.md:64]: Where do logs go? ALso, what is $Write_DIR shouldn't it be $Data_dir?

File: README.md [hunk 35]
```diff
@@ -1,7 +1,123 @@
...
+python eval_ff.py --path [path to ckpt folders] --dataset [V2/A/Sketch/R/V1]
+```
+If `dataset` flag is not passed, it will evaluate on standard ImageNet validation set. 
+
+### Evaluating MRL models 
+```
+python eval_MRL.py --path [path to weight checkpoint, need to be .pt file] --dataset [V2/A/Sketch/R/V1] (--efficient)
+```
+
+## [Evaluating pre-trained checkpoints](eval_ckpts/)
```
Comment (adityakusupati) [inline README.md]: pre-trained MRL checkpoints?

Comment (bhattg) [reply README.md]: Okay I think this is unintentionally pointing to w/o any training model. I meant already trained by us

File: README.md (hunk 34 above)
Comment (bhattg) [reply README.md:64]: There is a separate folder for train logs where we dump logs. Can change $WRITE_DIR to $DATA_DIR but it was done to make sure we have consistency with FFCV notations.

File: README.md [hunk 36]
```diff
@@ -6,44 +6,43 @@ This repository contains the code for training and evaluating Matryoshka Represe
 1. Set up
```
Comment (adityakusupati) [inline README.md:6]: Define MRL in line 4.

File: README.md [hunk 37]
```diff
@@ -6,44 +6,43 @@ This repository contains the code for training and evaluating Matryoshka Represe
...
+Pip install the requirements file in this directory:
 ```
 pip install -r requirements.txt
 ```
 We use PyTorch Distributed Data Parallel shared over 2 A100 GPUs and FFCV for the dataloader. FFCV uses 8 A100 GPUs, therefore we linearly downscale the learning rate by 4.
 
 ### Preparing the dataset.
-
-Generate ImageNet FFCV dataset from standard imagenet train set with the following command (`IMAGENET_DIR` should point to a PyTorch style [ImageNet dataset](https://github.com/MadryLab/pytorch-imagenet-dataset)):
+Download the standard Imagenet-1k train and val set, and store the validation set in root directory with folder named `val/`. Using same code snippet as [FFCV](https://github.com/libffcv/ffcv-imagenet), generate the dataset from standard imagenet train set with the following command (`IMAGENET_DIR` should point to a PyTorch style [ImageNet dataset](https://github.com/MadryLab/pytorch-imagenet-dataset)):
```
Comment (adityakusupati) [inline README.md]: ImageNet-1K

File: README.md [hunk 38]
```diff
@@ -52,8 +51,8 @@ fc_layer = MRL_Linear_Layer(nesting_list, num_classes=1000, efficient=efficient)
 ```
 Where, `nesting_list` is the list of dimensions we want to nest on, `num_classes` is the number of output feature and efficient flag if we want to do MRL-E.
 
-## Training ResNet50 models
-First of all change the directory to `train_and_eval/train`. While training, we dump  model ckpts, training logs and run command inside a folder. 
+## [Training ResNet50 models](train/)
+`cd train/`. While training, we dump  model ckpts, training logs and run command inside a folder. `$WRITE_DIR` is same variable that we used while creating the dataset.
```
Comment (adityakusupati) [inline README.md]: Add a line break after cd dtuff.

File: README.md [hunk 39]
```diff
@@ -88,24 +87,21 @@ By default, we start nesting from 8 dimensions (That is 2^3). In case we want to
...
+## [Inference on Trained Models](inference/)
 
-```
-python eval_ff.py --path [path to ckpt folders] --dataset [V2/A/Sketch/R/V1]
-```
-If `dataset` flag is not passed, it will evaluate on standard ImageNet validation set. 
+### Classification performance
+`cd inference/` . To evaluate the model, run the following command; argument in brackets are optional. To evaluate the Fixed Feature Baseline, just pass `--rep_size <dim>` to evaluate particular size dim. Script is also capable of using pytorch evaluation of standard imagenet-1k validation set (V1). To evaluate our uploaded checkpoints, please pass `--old_ckpt`. Our model checkpoints can be found [here](https://drive.google.com/drive/folders/1IEfJk4xp-sPEKvKn6eKAUzvoRV8ho2vq?usp=sharing).
```
Comment (adityakusupati) [inline README.md]: Add a line break. Full stops should not have space before them

File: README.md [hunk 40]
```diff
@@ -88,24 +87,21 @@ By default, we start nesting from 8 dimensions (That is 2^3). In case we want to
...
+python pytorch_eval.py --path <final_weigth.pt> --dataset <V2/A/Sketch/R/V1> [--tta] [--mrl] [--efficient] [--rep_size <dim>] [--old_ckpt]
 ```
 
-## [Evaluating pre-trained checkpoints](eval_ckpts/)
-We also provide path to the saved checkpoints which were used to generate the results in our paper. However, the MRL Linear layer had a slightly different re-factored [code](eval_ckpts/NestingLayer.py), therefore we have separate scripts to directly evaluate them. Saved model ckpts can be found [here](https://drive.google.com/drive/folders/1IEfJk4xp-sPEKvKn6eKAUzvoRV8ho2vq?usp=sharing). Please click [here](eval_ckpts/) to evaluate pre-trained checkpoints for MRL models. 
+In paper we only consider rep sizes in  `[8, 16, 32, 64, 128, 256, 512, 1024, 2048]`, therefore to evaluate on other rep. sizes change the variable `NESTING_LIST` in file `pytorch_eval.py`. 
+
+
+## [Model Analysis](model_analysis/)
+`cd model_analysis/` Here we provide 4 jupyter notebooks which contain performance visualization such as GradCAM images (for checkpoint models), superclass performance, model cascades and oracle upper bound. Please refer to its documentation [here](model_analysis/README.md).
```
Comment (adityakusupati) [inline README.md]: Add a line break.

File: inference/pytorch_eval.py [hunk 41]
```diff
@@ -0,0 +1,112 @@
+'''
```
Comment (aniketrege) [inline inference/pytorch_eval.py:1]: delete `.DS_Store` in this subdir

File: inference/imagenet_id.py [hunk 42]
```diff
@@ -0,0 +1,18 @@
+import numpy as np
```
Comment (aniketrege) [inline inference/imagenet_id.py:1]: Add a brief header comment explaining the use of this file

File: model_analysis/imagenet_id.py [hunk 43]
```diff
@@ -0,0 +1,18 @@
+import numpy as np
```
Comment (aniketrege) [inline model_analysis/imagenet_id.py:5]: Add header comment explaining the use of this file. How is this different from `~/inference/imagenet_id.py`?

File: train/write_imagenet.py [hunk 44]
```diff
@@ -0,0 +1,69 @@
+'''
```
Comment (aniketrege) [inline train/write_imagenet.py:1]: delete `.DS_Store` in this subdir

File: inference/pytorch_eval.py (hunk 41 above)
Comment (aniketrege) [reply inference/pytorch_eval.py:1]: also delete from root

File: model_analysis/README.md [hunk 45]
```diff
@@ -1,14 +1,5 @@
-## Evaluating pre-trained checkpoints
```
Comment (aniketrege) [inline model_analysis/README.md]: how is `imagenet_id.py` different from the file in `~/inference/` ?

**Review by aniketrege — CHANGES_REQUESTED** [review]
minor changes
